File: apb/health.py
"""apb.health — a zero-dependency HTTP health endpoint for PaaS deploys.

The bot itself is a long-polling **worker**: it never listens on a socket, it
only makes *outbound* HTTPS calls to ``api.telegram.org``.  That is a problem
for platforms that deploy "web services" and wait for a port to open before
declaring a release healthy::

    ==> Waiting for your service to be ready
    Deploy aborted - the new version was crash-looping.

This module runs a tiny threaded :mod:`http.server` on a background daemon
thread so those platforms have something to probe:

    GET /          → 200  short human-readable "ok"
    GET /health    → 200  JSON liveness  (up as soon as the process is up)
    GET /healthz   → 200  alias
    GET /live      → 200  alias
    GET /ready     → 200  JSON readiness (503 until the bot has logged in)
    GET /readyz    → 200/503 alias

Liveness and readiness are deliberately separate: ``/health`` answers 200 from
the moment the socket is bound (so a slow ``getMe`` round-trip to Telegram can
never be mistaken for a crash loop), while ``/ready`` flips to 200 only once
the bot is actually polling.

Nothing here is required to run the bot on your own machine — pass
``--no-health`` (or simply don't set ``PORT``/``APB_PORT``) and no socket is
opened at all.
"""

from __future__ import annotations

import json
import logging
import os
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

log = logging.getLogger(__name__)

LIVE_PATHS = frozenset(("/", "/health", "/healthz", "/live", "/liveness"))
READY_PATHS = frozenset(("/ready", "/readyz", "/readiness"))
DEFAULT_HOST = "0.0.0.0"          # PaaS health checks come from outside the box
DEFAULT_PORT = 8080


def resolve_port(cli_port=None):
    """Pick the health-check port.

    Order: ``--port`` → ``$APB_PORT`` → ``$PORT`` (injected by Heroku, Railway,
    Render, Fly, Koyeb, Google Cloud Run …) → ``None`` (no health server).
    """
    for raw in (cli_port, os.environ.get("APB_PORT"), os.environ.get("PORT")):
        if raw is None:
            continue
        raw = str(raw).strip()
        if raw.isdigit() and 0 < int(raw) < 65536:
            return int(raw)
        if raw:
            log.warning("ignoring unusable port %r", raw)
    return None


class _Handler(BaseHTTPRequestHandler):
    """Minimal GET/HEAD handler — always sends Content-Length (HTTP/1.1)."""

    protocol_version = "HTTP/1.1"
    server_version = "AdvancedPostingBot/1.0"
    health = None                                   # set by HealthServer

    # ------------------------------------------------------------ plumbing
    def log_message(self, fmt, *args):              # keep the bot's log clean
        log.debug("health %s", fmt % args)

    def _send(self, code, body, ctype="application/json; charset=utf-8"):
        if isinstance(body, str):
            body = body.encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", ctype)
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        if self.command != "HEAD":
            self.wfile.write(body)

    def _json(self, code, payload):
        self._send(code, json.dumps(payload, ensure_ascii=False, indent=2) + "\n")

    # -------------------------------------------------------------- routes
    def do_GET(self):
        self._route()

    def do_HEAD(self):
        self._route()

    def _route(self):
        path = (self.path or "/").split("?", 1)[0].rstrip("/") or "/"
        hs = self.health
        if hs is not None:
            hs.hits += 1
        state = hs.snapshot() if hs else {"status": "unknown"}

        if path in LIVE_PATHS:
            if path == "/":
                self._send(200, "Advanced Posting Bot — ok\n"
                                "  liveness : /health\n"
                                "  readiness: /ready\n",
                           ctype="text/plain; charset=utf-8")
            else:
                self._json(200, state)
        elif path in READY_PATHS:
            ready = bool(state.get("ready"))
            self._json(200 if ready else 503, state)
        else:
            self._json(404, {"status": "not_found", "path": path,
                             "hint": "try /health or /ready"})


class HealthServer(threading.Thread):
    """Background daemon thread serving the health endpoints above."""

    def __init__(self, port=None, host=DEFAULT_HOST, stats=None):
        super().__init__(daemon=True, name="apb-health")
        self.host = host
        self.port = port if port is not None else DEFAULT_PORT
        self.stats = stats                  # optional callable → dict of extras
        self.started_at = time.time()
        self.hits = 0
        self._ready = False
        self._ready_detail = "starting"
        self._lock = threading.Lock()
        self._httpd = None
        self._bound = False
        self._bound_evt = threading.Event()

    # ------------------------------------------------------------- lifecycle
    def start(self, wait=5.0):
        """Start the thread and block until the port is actually listening.

        Platforms begin probing the moment the container starts, and the caller
        wants to know whether the bind worked before printing the banner — so
        this waits for the socket (or for the bind to fail) instead of
        returning the instant the thread is spawned.
        """
        super().start()
        self._bound_evt.wait(wait)
        return self._bound

    def run(self):
        _Handler.health = self
        try:
            self._httpd = ThreadingHTTPServer((self.host, self.port), _Handler)
        except OSError as exc:
            log.error("health server could not bind %s:%s — %s "
                      "(the bot keeps running; deploy platforms will see no port)",
                      self.host, self.port, exc)
            self._bound_evt.set()
            return
        self._httpd.daemon_threads = True
        self._bound = True
        self._bound_evt.set()
        log.info("health server listening on http://%s:%s/health",
                 self.host, self.bound_port)
        try:
            self._httpd.serve_forever(poll_interval=0.5)
        finally:
            try:
                self._httpd.server_close()
            except OSError:
                pass

    def stop(self):
        if self._httpd is not None:
            self._httpd.shutdown()
            self._httpd.server_close()
            self._httpd = None
        self._bound = False

    def mark_ready(self, detail="polling"):
        with self._lock:
            self._ready = True
            self._ready_detail = detail
        log.info("health: ready (%s)", detail)

    def mark_not_ready(self, detail="degraded"):
        with self._lock:
            self._ready = False
            self._ready_detail = detail

    @property
    def bound(self):
        return self._bound

    @property
    def bound_port(self):
        """Actual listening port (``port=0`` → whatever the OS assigned)."""
        if self._httpd is not None:
            try:
                return self._httpd.server_address[1]
            except (AttributeError, IndexError):
                pass
        return self.port

    @property
    def url(self):
        shown = "localhost" if self.host in ("0.0.0.0", "::") else self.host
        return "http://{}:{}/health".format(shown, self.bound_port)

    # ---------------------------------------------------------------- status
    def snapshot(self):
        with self._lock:
            ready, detail = self._ready, self._ready_detail
        out = {
            "status": "ok" if ready else "starting",
            "service": "advanced-posting-bot",
            "runtime": "python",
            "ready": ready,
            "detail": detail,
            "uptime_s": round(time.time() - self.started_at, 1),
            "pid": os.getpid(),
            "health_hits": self.hits,
        }
        if self.stats:
            try:
                extra = self.stats()
                if isinstance(extra, dict):
                    out.update(extra)
            except Exception as exc:                       # never break /health
                out["stats_error"] = str(exc)
        return out

File: src/health.js
/**
 * src/health.js — the HTTP face of the bot.
 *
 * The bot itself is a long-polling **worker**: it only makes outbound HTTPS
 * calls to api.telegram.org.  Platforms that run "web services", however,
 * wait for a port to open before they declare a release healthy:
 *
 *     ==> Waiting for your service to be ready
 *     Deploy aborted - the new version was crash-looping.
 *
 * So this module serves a tiny, dependency-free HTTP server on $PORT:
 *
 *   GET /                 → 200  status page (HTML in a browser, text otherwise)
 *   GET /health /healthz  → 200  JSON liveness (up as soon as the socket is up)
 *   GET /live /liveness   → 200  alias
 *   GET /ready /readyz    → 200 JSON readiness (503 until the bot has logged in)
 *   GET /status           → 200  JSON, same body as /health
 *
 * Liveness and readiness are deliberately separate: /health answers 200 from
 * the instant the socket is bound (so a slow getMe round-trip can never be
 * mistaken for a crash loop) while /ready flips to 200 only once polling.
 *
 * Deliberately no X-Frame-Options / CSP headers: the status page is meant to
 * be embeddable in hosting dashboards (Veroa, Railway …) and previews.
 */

import http from 'node:http';
import os from 'node:os';
import { logger } from './logger.js';

const log = logger('health');

export const LIVE_PATHS = new Set(['/', '/health', '/healthz', '/live', '/liveness', '/status']);
export const READY_PATHS = new Set(['/ready', '/readyz', '/readiness']);

/** cliPort → $APB_PORT → $PORT → default. Returns null when disabled. */
export function resolvePort(cliPort = null, fallback = null) {
  const candidates = [cliPort, process.env.APB_PORT, process.env.PORT];
  for (const raw of candidates) {
    if (raw === null || raw === undefined || raw === '') continue;
    const value = String(raw).trim();
    if (/^\d+$/.test(value) && Number(value) > 0 && Number(value) < 65536) return Number(value);
    log.warn(`ignoring unusable port ${JSON.stringify(raw)}`);
  }
  return fallback;
}

function esc(value) {
  return String(value).replace(/[&<>"']/g, (c) => ({
    '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;',
  }[c]));
}

export class HealthServer {
  constructor({ port, host = '0.0.0.0', stats = null } = {}) {
    this.host = host;
    this.port = port;
    this.stats = stats;              // optional () => ({...extra})
    this.startedAt = Date.now();
    this.hits = 0;
    this.ready = false;
    this.readyDetail = 'starting';
    this.server = null;
    this.boundPort = port;
  }

  /** Start listening; resolves with `{ ok, port, error }`. */
  start() {
    return new Promise((resolve) => {
      const server = http.createServer((req, res) => this.#handle(req, res));
      server.keepAliveTimeout = 65000;
      server.headersTimeout = 70000;
      const done = (ok, error = null) => resolve({ ok, port: this.boundPort, error });
      server.on('error', (err) => {
        log.error(`health server could not bind ${this.host}:${this.port} — ${err.message} `
          + '(the bot keeps running; a port-probing platform will see no port)');
        done(false, err);
      });
      server.listen(this.port, this.host, () => {
        this.server = server;
        this.boundPort = server.address().port;
        log.info(`listening on http://${this.host}:${this.boundPort}  (/health · /ready)`);
        done(true);
      });
      this.server = server;
    });
  }

  stop() {
    if (!this.server) return Promise.resolve();
    const server = this.server;
    this.server = null;
    return new Promise((resolve) => {
      try {
        server.close(() => resolve());
        server.closeAllConnections?.();
      } catch {
        resolve();
      }
      setTimeout(resolve, 1500).unref?.();
    });
  }

  markReady(detail = 'polling') {
    this.ready = true;
    this.readyDetail = detail;
    log.info(`ready (${detail})`);
  }

  markNotReady(detail = 'degraded') {
    this.ready = false;
    this.readyDetail = detail;
  }

  get url() {
    const host = ['0.0.0.0', '::'].includes(this.host) ? 'localhost' : this.host;
    return `http://${host}:${this.boundPort}/health`;
  }

  snapshot() {
    const out = {
      status: this.ready ? 'ok' : 'starting',
      service: 'advanced-posting-bot',
      runtime: 'node',
      node: process.version,
      ready: this.ready,
      detail: this.readyDetail,
      uptime_s: Math.round(((Date.now() - this.startedAt) / 1000) * 10) / 10,
      pid: process.pid,
      memory_mb: Math.round((process.memoryUsage().rss / 1048576) * 10) / 10,
      health_hits: this.hits,
      host: os.hostname(),
    };
    if (this.stats) {
      try {
        const extra = this.stats();
        if (extra && typeof extra === 'object') Object.assign(out, extra);
      } catch (err) {
        out.stats_error = String(err?.message || err);
      }
    }
    return out;
  }

  #statusPage(state) {
    const rows = Object.entries(state).map(([k, v]) => (
      `<tr><th>${esc(k)}</th><td>${esc(typeof v === 'object' ? JSON.stringify(v) : v)}</td></tr>`
    )).join('');
    return `<!doctype html>
<html lang="en"><head><meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>Advanced Posting Bot — status</title>
<style>
 :root{color-scheme:dark light}
 body{font:15px/1.5 ui-sans-serif,system-ui,-apple-system,Segoe UI,Roboto,sans-serif;
      margin:0;padding:32px;background:#0f1115;color:#e8eaf0}
 .card{max-width:640px;margin:0 auto;background:#171a21;border:1px solid #262b36;
       border-radius:14px;padding:24px}
 h1{margin:0 0 4px;font-size:20px}
 .pill{display:inline-block;padding:3px 10px;border-radius:999px;font-size:12px;font-weight:600}
 .ok{background:#10361f;color:#4ade80}.wait{background:#3a2a10;color:#fbbf24}
 table{width:100%;border-collapse:collapse;margin-top:18px;font-size:13.5px}
 th{text-align:left;color:#98a2b3;font-weight:500;white-space:nowrap;padding:5px 12px 5px 0}
 td{padding:5px 0;word-break:break-word}
 code{background:#0f1115;padding:2px 6px;border-radius:6px}
 a{color:#60a5fa}
</style></head><body><div class="card">
<h1>🤖 Advanced Posting Bot</h1>
<p><span class="pill ${state.ready ? 'ok' : 'wait'}">${state.ready ? 'ready' : 'starting / not logged in'}</span></p>
<p style="color:#98a2b3">${esc(state.detail || '')}</p>
<table>${rows}</table>
<p style="margin-top:18px;color:#98a2b3;font-size:13px">
 Health: <a href="/health"><code>/health</code></a> ·
 Readiness: <a href="/ready"><code>/ready</code></a> ·
 JSON: <a href="/status"><code>/status</code></a>
</p>
</div></body></html>`;
  }

  #handle(req, res) {
    try {
      this.hits += 1;
      const url = new URL(req.url || '/', 'http://localhost');
      let path = url.pathname.replace(/\/+$/, '');
      if (!path) path = '/';
      const state = this.snapshot();
      const headers = { 'cache-control': 'no-store', connection: 'close' };

      const send = (code, body, contentType) => {
        const payload = Buffer.from(body, 'utf8');
        res.writeHead(code, { ...headers, 'content-type': contentType, 'content-length': payload.length });
        res.end(req.method === 'HEAD' ? undefined : payload);
      };

      if (READY_PATHS.has(path)) {
        send(this.ready ? 200 : 503, `${JSON.stringify(state, null, 2)}\n`, 'application/json; charset=utf-8');
        return;
      }
      if (path === '/' || path === '/health' || path === '/healthz' || path === '/live' || path === '/liveness') {
        const wantsHtml = String(req.headers.accept || '').includes('text/html');
        if (path === '/' && wantsHtml) send(200, this.#statusPage(state), 'text/html; charset=utf-8');
        else send(200, `${JSON.stringify(state, null, 2)}\n`, 'application/json; charset=utf-8');
        return;
      }
      if (path === '/status') {
        send(200, `${JSON.stringify(state, null, 2)}\n`, 'application/json; charset=utf-8');
        return;
      }
      if (path === '/favicon.ico') {
        send(204, '', 'image/x-icon');
        return;
      }
      send(404, `${JSON.stringify({ status: 'not_found', path, hint: 'try /health or /ready' }, null, 2)}\n`, 'application/json; charset=utf-8');
    } catch (err) {
      log.warn(`health handler error: ${err?.message || err}`);
      try {
        res.writeHead(500, { 'content-type': 'text/plain' });
        res.end('error\n');
      } catch { /* socket already gone */ }
    }
  }
}

export default HealthServer;
